### PR TITLE
Message drafts

### LINKED ISSUE #2785: Remember user prompt when switching between chats
Currently, The prompt gets lost when we type something in the chat form and switch to another chat and come back to the chat we were editing.

[remember-prompt.webm](https://user-images.githubusercontent.com/94373953/233428793-31fadbc0-a712-499d-a30a-6bfb2eeaa57f.webm)

It would be a very good UX if we remember the prompt for individual chat.

### LINKED ISSUE #2931: Message ranking In Conversations
This is a proposal for multiple messages to be generated per prompt to collect additional RLHF data from user conversations. It would work by generating the first 'x' tokens of a message and then prompting the user to select one which would be generated to completion. The value of 'x' would change based on the load on the inference servers at the moment ranging from whole messages to completely disabling the feature under high load. This would give lots of ranking data for RLHF while not butchering the performance of the site, if the server is at low usage then you get lots of RLHF data while improving quality of life for users, if the server is at capacity then it is simply disabled. Giving the option to see 'x' number of tokens from a few possible answers would also reduce the number of regenerations reducing the net increase in load.

---

It would go something like this:

User: What is a croissant?

1. A croissant is a type of French pastry made with layers of yeast-raised dough that are rolled and folded several times. The resulting dough is shaped into an crescent shape and baked until flaky, crisp, and...

2. Croissants are buttery and delicious. They originated in Austria during the 13th century, with the first recorded use of the word “croissant” appearing in a French manuscript from the year...

3. A crescent shaped bread rolls often eaten at breakfast. (Sometimes short answers can fit within 'x' tokens)

(User selects reply 1)

Assistant: "A croissant is a type of French pastry made with layers of yeast-raised dough that are rolled and folded several times. The resulting dough is shaped into an crescent shape and baked until flaky, crisp, and golden brown in color.

Traditionally, croissants were made with butter, which gave them their signature rich flavor and flakiness. However, modern versions may also use margarine or other spreads to reduce the fat content. Croissants can be served plain as a breakfast bread or filled with various ingredients such as jam, cheese, or chocolate for added sweetness or savory flavors.

Croissants have become popular around the world, particularly in countries with strong ties to France such as Belgium, Switzerland, and Canada, as well as many other parts of Europe and North America where French cuisine has gained a following. They are often served at cafes alongside coffee or tea, making it a staple breakfast item."

### PR DESCRIPTION
**Pull request by someone13574**
closes #2931 (slightly changed goal based on advice from the discord, generate full messages, not 'x' tokens. Full messages are more useful data)

- [x] Create draft selection UI
- [x] Draft inference
- [x] Option to regenerate drafts and serve 3 new ones
- [x] Remember last viewed sibling message
- [x] Store selected draft training data for RLHF
- ~~[ ] Disable drafts when queue is too long / server is under load~~ (Suggested to leave to next PR in the discord)
- [x] Draft markdown rendering
- [x] 'Used plugin' UI for drafts
- [x] Resolve merge conflicts

### CODE REVIEW

File: inference/server/alembic/versions/2023_05_07_0923-19178d144265_add_sibling_active_to_message.py
Comment (olliestanley) [inline inference/server/alembic/versions/2023_05_07_0923-19178d144265_add_sibling_active_to_message.py:1]: Can we merge both Alembic scripts into one before merging this? Will help reduce clutter

File: inference/server/oasst_inference_server/models/chat.py
Comment (olliestanley) [inline inference/server/oasst_inference_server/models/chat.py:22]: What does `active_sibling` mean exactly? Feels like an odd name

File: inference/server/oasst_inference_server/models/chat.py
Comment (olliestanley) [inline inference/server/oasst_inference_server/models/chat.py:43]: I guess that the way this works is that for each prompter message with multiple drafted replies, each reply has a list of message IDs which are "inferior" to it?

File: website/src/components/Chat/ChatConversation.tsx
Comment (olliestanley) [inline website/src/components/Chat/ChatConversation.tsx:124]: I wonder if this will be a problem when plugins are enabled... With sufficient retries/depth plugins take up a lot of model time/compute already, several drafts might be a struggle. Maybe we could reduce plugin max depth for drafts?

**Review by olliestanley — CHANGES_REQUESTED** [review]

**Comment by olliestanley** [comment]
Thanks, I think this mostly makes sense, I left a couple of questions and one change request

File: inference/server/oasst_inference_server/models/chat.py
Comment (someone13574) [reply inference/server/oasst_inference_server/models/chat.py:22]: In a conversation tree you can have messages which you can switch between using the arrows at the bottom. This variable just stores whether or not that particular message was the one which was last showed. The 'sibling' part is referring to how these messages come from the same parent message so I thought it would be a good term to use. If you have a better alternative I can change it.

File: inference/server/oasst_inference_server/models/chat.py
Comment (someone13574) [reply inference/server/oasst_inference_server/models/chat.py:43]: Yes, when the user selects a draft message the two other messages have their ID's added to the selected message's inferior message list.

File: website/src/components/Chat/ChatConversation.tsx
Comment (someone13574) [reply website/src/components/Chat/ChatConversation.tsx:124]: I might just disable drafts when plugins are enabled until the disabling under load feature is added as to not overload the inference server. Reducing max iterations is also a possibility.

Comment (olliestanley) [reply website/src/components/Chat/ChatConversation.tsx:124]: > I might just disable drafts when plugins are enabled until the disabling under load feature is added as to not overload the inference server. Reducing max iterations is also a possibility.

Yes, not only overload but also just taking a lot of time for the user to get the drafts... Maybe include a setting to toggle enabling this for plugins, and have it disabled by default

Comment (someone13574) [reply website/src/components/Chat/ChatConversation.tsx:126]: I've added an ENABLE_DRAFTS_FOR_PLUGINS environment variable which is defaulted to false

File: inference/server/oasst_inference_server/models/chat.py
Comment (olliestanley) [reply inference/server/oasst_inference_server/models/chat.py:22]: Hmm ok. This feels like it is really information about the chat, not the message. Can we not just store a single `active_message` value in the chat table, which is the final message of the currently active path down the chat tree? Then from this it is already implied what siblings are active earlier in the tree

This would also simplify the endpoint, we could merge it into `update_chat` endpoint

File: inference/server/oasst_inference_server/models/chat.py
Comment (olliestanley) [reply inference/server/oasst_inference_server/models/chat.py:43]: Ok. Thinking from the perspective of ease of querying/interpreting this data. We could probably have a `draft_evaluation` table with fields like:
```
id: str
user_id: str
selected_message_id: str
inferior_message_ids: list[str]
```

File: inference/server/oasst_inference_server/user_chat_repository.py
Comment (olliestanley) [inline inference/server/oasst_inference_server/user_chat_repository.py:274]: We can probably remove the refresh and return here

**Review by olliestanley — COMMENTED** [review]
Thanks! This seems fine to me from backend perspective now. Will approve tomorrow if neither of Yannic/Andreas wish to review first

File: inference/server/oasst_inference_server/models/chat.py
Comment (someone13574) [reply inference/server/oasst_inference_server/models/chat.py]: Would it make sense to rename the `draft_evaluation` table to a more general `message_evaluation` so that it could be reused in the future for something like ChatGPT's "Better, Worse, Same" on message regeneration?

Comment (olliestanley) [reply inference/server/oasst_inference_server/models/chat.py]: Yeah this could make sense - will leave up to you

**Review by olliestanley — APPROVED** [review]
Backend changes LGTM. Have not reviewed website changes.

File: website/src/components/Chat/ChatAssistantDraftViewer.tsx
Comment (notmd) [inline website/src/components/Chat/ChatAssistantDraftViewer.tsx:94]: Is there any reason you use a key here?

File: website/src/components/Chat/ChatAssistantDraftViewer.tsx
Comment (notmd) [inline website/src/components/Chat/ChatAssistantDraftViewer.tsx:34]: This component is way too big, we should extract it to 2 component. One for render the content, and one for display the paging logic.

File: website/src/components/Chat/ChatAssistantDraftViewer.tsx
Comment (notmd) [inline website/src/components/Chat/ChatAssistantDraftViewer.tsx:129]: TS is warning that the message is any, and will be error if we turn on strict mode. Extract to sub-component will make the code safer here.

File: website/src/components/Chat/ChatConversation.tsx
Comment (notmd) [inline website/src/components/Chat/ChatConversation.tsx:171]: This line causes a lot of race condition, and make the queue info flashing all the times. Can we call setQueueInfo only 1 time or move the queue info to inside the message?

File: website/src/components/Chat/ChatConversationTree.tsx
Comment (notmd) [inline website/src/components/Chat/ChatConversationTree.tsx:58]: Why do we need this?

File: website/src/components/Chat/ChatConversation.tsx
Comment (notmd) [inline website/src/components/Chat/ChatConversation.tsx:404]: Can you add typing to this function?

File: website/src/components/Chat/ChatConversationTree.tsx
Comment (notmd) [inline website/src/components/Chat/ChatConversationTree.tsx:61]:
```suggestion
  const [index, setIndex] = useState(0);
```

**Review by notmd — COMMENTED** [review]
Thank you. Look good overall I might need to take another round, there are some eslint warnings about the hook deps, could you fix that? Also, there are many duplicated codes, I'm fine with that, we could refactor later.

File: website/public/locales/en/chat.json
Comment (notmd) [inline website/public/locales/en/chat.json:41]: Can you move this to the common file?

**Comment by notmd** [comment]
Also, can we make an option to allow users to choose whether they want to generate drafts or not?

File: website/src/components/Chat/ChatAssistantDraftViewer.tsx
Comment (notmd) [inline website/src/components/Chat/ChatAssistantDraftViewer.tsx:78]:
```suggestion
              key={`draft-message-${index}`}
```

**Review by notmd — APPROVED** [review]
Thank you, need 1 more approval from Yannic or Andreas.

File: website/src/components/Chat/ChatConversation.tsx
Comment (andreaskoepf) [inline website/src/components/Chat/ChatConversation.tsx:225]: This string should be added to `public/locales/en/chat.json` for proper app localization.

File: website/src/types/Config.ts
Comment (andreaskoepf) [inline website/src/types/Config.ts:3]: Since the meaning of this configuration option is not immediately clear (at least it was not fro me quickly reading through the PR) it would be good to have a comment here which explains what it does.

File: inference/server/oasst_inference_server/models/chat.py
Comment (andreaskoepf) [reply inference/server/oasst_inference_server/models/chat.py]: Sorry only seeing this now: How are the multiple generated draft messages now referenced in the database? And was is finally stored in the chat-message tree?

File: inference/server/alembic/versions/2023_05_18_1748-2d67fbdc5b46_add_active_messsage_id_and_message_eval.py
Comment (andreaskoepf) [inline inference/server/alembic/versions/2023_05_18_1748-2d67fbdc5b46_add_active_messsage_id_and_message_eval.py:45]: Why is this a JSONB column?

**Review by andreaskoepf — CHANGES_REQUESTED** [review]
Thanks a lot for this important feature.
I would appreciate feedback to the following points:

1. It seems as if the draft generation would always be active when `NUM_GENERATED_DRAFTS > 1`. This will completely break any regular chat-experience and force the user to go through the message-selection process, even for simple "Hello" prompts. In general we definitely want to collect this form of feedback data but the question is if we really should make regular use of assistant impossible. If message alternatives would be generated when user presses regenerate or down-vote it would IMO be a much more unobtrusive experience and users would get additional option when they actually need them.

2. It is not clear to me how draft messages are tracked now in the database. Could you please explain the meaning the `active_message_id` column in `chat` table (and why is it a json field)?  How are drafts messages stored and referenced in the database?

**Comment by someone13574** [comment]
> Thanks a lot for this important feature. I would appreciate feedback to the following points:
> 
>     1. It seems as if the draft generation would always be active when `NUM_GENERATED_DRAFTS > 1`. This will completely break any regular chat-experience and force the user to go through the message-selection process, even for simple "Hello" prompts. In general we definitely want to collect this form of feedback data but the question is if we really should make regular use of assistant impossible. If message alternatives would be generated when user presses regenerate or down-vote it would IMO be a much more unobtrusive experience and users would get additional option when they actually need them.
> 
>     2. It is not clear to me how draft messages are tracked now in the database. Could you please explain the meaning the `active_message_id` column in `chat` table (and why is it a json field)?  How are drafts messages stored and referenced in the database?

1. I'll make the drafts less intrusive by only generating drafts if either the prompt is long (thus indicating a more complex request) or if the user regenerates a message (indicating that additional options would be beneficial). This should make the process less intrusive and act as a filter against data we really don't need comparisons for (trivial tasks or general conversation).
2. Draft messages are stored the same way as normal message and the comparisons are stored in an new table called `message_eval` where it has the id of the selected message and the id's of the messages it was selected over. The `active_message_id` stores the id of the last message in the currently visible thread so that the conversation can be loaded in the same state as it was when a user closes it. This is because all the unselected drafts are also added to the conversation as other threads and the last message for a given parent message wont always be the one the user was actually using, so loading onto that unused message instead of where the user actually was writing was very bad for ux, so I made it remember which thread was last selected. I'd be open to a better name. Also using a json was a mistake, I'll fix it.

File: inference/server/oasst_inference_server/models/chat.py
Comment (someone13574) [reply inference/server/oasst_inference_server/models/chat.py]: The draft messages are stored as regular messages in the messages table. The message comparisons are stored in a new table called `message_eval` which contains the chat id, user id, message id of the selected message, and the message ids it was selected over.

**Review by andreaskoepf — APPROVED** [review]

**Pull request merged**
